functions/go/project-service-list/README.md

```diff
@@ -0,0 +1,117 @@
+    * Sets namespace if any defined for `ProjectServiceList` custom resource to each generated resource.
+    * Sets projectID if any defined for `ProjectServiceList` custom resource to each generated resource.
+1. Each generated [project service](https://cloud.google.com/config-connector/docs/reference/resource-docs/serviceusage/service) resource
+has a `blueprints.cloud.google.com/managed-by-project-services-list` annotation. This annotation allows `project-services-list` function to
+track generated resources for the declarative management of the generated resources. Any changes made to the generate resources will be overwritten and should be made to the `ProjectServiceList` CRD instead.
+
+### FunctionConfig
+
+This function only supports a custom resource `functionConfig` of kind `ProjectServiceList`.
+
+#### `ProjectServiceList`
+
+A functionConfig of kind `ProjectServiceList` has the following supported parameters:
+
+```yaml
+apiVersion: blueprints.cloud.google.com/v1alpha1
+kind: ProjectServiceList
+metadata:
+  name: my-project-services
+  annotations:
+    cnrm.cloud.google.com/deletion-policy: false
+spec:
+  services: # list of services to generate
+    - compute.googleapis.com
+  projectID: foo # explicit project ID to be set for the generated services
+```
+
+| Field        |  Description | Example
+| -----------: |  ----------- | -----------
+`spec.services[]`    | A list of GCP services to enable | [compute.googleapis.com,bigquery.googleapis.com]
+`spec.projectID`   | Project ID where the services should be enabled. | my-project-id
+
+<!--mdtogo-->
+
+## Examples
+
+<!--mdtogo:Examples-->
+
+Let's start with the `functionConfig` for enabling two services `compute.googleapis.com` and `redis.googleapis.com` in a GCP Project `proj1`.
+
+```yaml
+# services-config.yaml
+apiVersion: blueprints.cloud.google.com/v1alpha1
+kind: ProjectServiceList
+metadata:
+  name: proj1-service
+spec:
+  services:
+    - compute.googleapis.com
+    - redis.googleapis.com
+  projectID: proj1
+```
+
+Invoke the function:
+
+```shell
+kpt fn eval --image gcr.io/kpt-fn/project-services-list:unstable --fn-config /tmp/services-config.yaml
+```
+
+Generated resources looks like the following:
+
+```yaml
+# service_proj1-service-compute.yaml
+apiVersion: serviceusage.cnrm.cloud.google.com/v1beta1
+kind: Service
+metadata:
+  name: proj1-service-compute
+  annotations:
+    blueprints.cloud.google.com/managed-by-project-services-list: 'proj1-service'
+spec:
+  resourceID: compute.googleapis.com
+  projectRef:
+    external: proj1
+```
+
+```yaml
+# service_proj1-service-redis.yaml
+apiVersion: serviceusage.cnrm.cloud.google.com/v1beta1
+kind: Service
+metadata:
+  name: proj1-service-redis
+  annotations:
+    blueprints.cloud.google.com/managed-by-project-services-list: 'proj1-service'
+spec:
+  resourceID: redis.googleapis.com
+  projectRef:
+    external: proj1
+```
```

Comment:
A few more questions:

1. Can the function scan resources in the package and generate the resources for enabling the GCP APIs automatically ? (may be it does, it wasn't clear from the documentation).
2. Ordering: so when do `kpt live apply`, it is expected to apply the resource that enable the API for a service first, then the instance for the service ?

Reply:
1) No but generate services #554 does this. This requires an explicit list, was https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/5432cdca9675ed6c7c2fbab13fc916ece0b3d2dc/functions/go/project-service-list/README.md#usage clear in this or should I add some more verbiage. We can also cross link to generate-services.

2) kpt will apply both together. Resource will fail at first, API will succeed and eventually resource will succeed. We can make this better when we have dependency ordering in kpt.